src/main/java/de/tum/in/www1/artemis/service/iris/session/IrisChatSessionService.java

```diff
@@ -150,105 +166,100 @@ public void checkRateLimit(User user) {
+        var participations = programmingExerciseStudentParticipationRepository.findAllWithSubmissionsByExerciseIdAndStudentLogin(exercise.getId(), user.getLogin());
+        if (participations.isEmpty()) {
+            return null;
         }
-        if (studentParticipations.isEmpty()) {
+        return participations.get(participations.size() - 1).getSubmissions().stream().max(Submission::compareTo)
+                .flatMap(sub -> programmingSubmissionRepository.findWithEagerBuildLogEntriesById(sub.getId())).orElse(null);
+    }
+
+    private Repository templateRepository(ProgrammingExercise exercise) {
+        return templateProgrammingExerciseParticipationRepository.findByProgrammingExerciseId(exercise.getId()).map(participation -> {
             try {
-                templateRepo = gitService.getOrCheckoutRepository(templateParticipation.get().getVcsRepositoryUri(), true);
+                return gitService.getOrCheckoutRepository(participation.getVcsRepositoryUri(), true);
             }
             catch (GitAPIException e) {
-                throw new InternalServerErrorException("Iris cannot function without template participation");
+                return null;
             }
-            parameters.put("templateRepository", repositoryService.getFilesWithContent(templateRepo));
-            return;
-        }
+        }).orElseThrow(() -> new InternalServerErrorException("Iris cannot function without template participation"));
+    }
 
-        try {
-            templateRepo = gitService.getOrCheckoutRepository(templateParticipation.get().getVcsRepositoryUri(), true);
-            studentRepo = gitService.getOrCheckoutRepository(studentParticipations.get(studentParticipations.size() - 1).getVcsRepositoryUri(), true);
-        }
-        catch (GitAPIException e) {
-            throw new InternalServerErrorException("Could not fetch existing student or template participation");
-        }
-        parameters.put("templateRepository", repositoryService.getFilesWithContent(templateRepo));
-        parameters.put("studentRepository", repositoryService.getFilesWithContent(studentRepo));
+    private Optional<Repository> studentRepository(@Nullable ProgrammingSubmission latestSubmission) {
+        return Optional.ofNullable(latestSubmission).map(sub -> (ProgrammingExerciseParticipation) sub.getParticipation()).map(participation -> {
+            try {
+                return gitService.getOrCheckoutRepository(participation.getVcsRepositoryUri(), true);
+            }
+            catch (GitAPIException e) {
+                log.error("Could not fetch existing student participation repository", e);
+                return null;
+            }
+        });
+    }
 
-        var oldTreeParser = new FileTreeIterator(templateRepo);
-        var newTreeParser = new FileTreeIterator(studentRepo);
+    private String getGitDiff(Repository from, Repository to) {
```

Comment:
Not for this PR necessarily, but this should ideally be put into the GitDiffService, so that all the git diffing logic is in one place for easy reusability.